server/src/main/java/org/elasticsearch/action/datastreams/autosharding/DataStreamAutoShardingService.java

```diff
@@ -0,0 +1,473 @@
+        "data_streams.auto_sharding.increase_shards.cooldown",
+        TimeValue.timeValueMinutes(15),
+        TimeValue.timeValueSeconds(0),
+        Setting.Property.Dynamic,
+        Setting.Property.NodeScope
+    );
+
+    /**
+     * Represents the minimum amount of time between two scaling events if the next event will reduce the number of shards.
+     */
+    public static final Setting<TimeValue> DATA_STREAMS_AUTO_SHARDING_DECREASE_SHARDS_COOLDOWN = Setting.timeSetting(
+        "data_streams.auto_sharding.decrease_shards.cooldown",
+        TimeValue.timeValueDays(3),
+        TimeValue.timeValueSeconds(0),
+        Setting.Property.Dynamic,
+        Setting.Property.NodeScope
+    );
+
+    /**
+     * Represents the minimum number of write threads we expect to have in the environments where auto sharding will be enabled.
+     */
+    public static final Setting<Integer> CLUSTER_AUTO_SHARDING_MIN_NUMBER_WRITE_THREADS = Setting.intSetting(
+        "cluster.auto_sharding.min_number_of_write_threads",
+        2,
+        1,
+        Setting.Property.Dynamic,
+        Setting.Property.NodeScope
+    );
+
+    /**
+     * Represents the maximum number of write threads we expect to have in the environments where auto sharding will be enabled.
+     */
+    public static final Setting<Integer> CLUSTER_AUTO_SHARDING_MAX_NUMBER_WRITE_THREADS = Setting.intSetting(
+        "cluster.auto_sharding.max_number_of_write_threads",
+        32,
+        1,
+        Setting.Property.Dynamic,
+        Setting.Property.NodeScope
+    );
+    private final ClusterService clusterService;
+    private final boolean isAutoShardingEnabled;
+    private final FeatureService featureService;
+    private final LongSupplier nowSupplier;
+    private volatile TimeValue increaseShardsCooldown;
+    private volatile TimeValue reduceShardsCooldown;
+    private volatile int minNumberWriteThreads;
+    private volatile int maxNumberWriteThreads;
+    private volatile List<String> dataStreamExcludePatterns;
+
+    public enum AutoShardingType {
+        INCREASE_NUMBER_OF_SHARDS,
+        DECREASES_NUMBER_OF_SHARDS,
+        NO_CHANGE_REQUIRED,
+        NOT_APPLICABLE
+    }
+
+    /**
+     * Represents an auto sharding recommendation. It includes the current and target number of shards together with a remaining cooldown
+     * period that needs to lapse before the current recommendation should be applied.
+     * <p>
+     * If auto sharding is not applicable for a data stream (e.g. due to {@link #DATA_STREAMS_AUTO_SHARDING_EXCLUDES_SETTING}) the target
+     * number of shards will be 0 and cool down remaining {@link TimeValue#MAX_VALUE}.
+     */
+    public record AutoShardingResult(
+        AutoShardingType type,
+        int currentNumberOfShards,
+        int targetNumberOfShards,
+        TimeValue coolDownRemaining,
+        @Nullable Double writeLoad
+    ) implements Writeable, ToXContentObject {
```

Comment:
We probably also want to have a test class for this class that extends `AbstractXContentSerializingTestCase`

Reply:
Ah yes, great catch. Due to the complications of this being used in a rollover Condition (the parsing is different when parsed independently vs via RolloverInfo - I've cried next to the parsing code for a full day to realize that an independent Condition being a named writeable will start parsing _inside_ the object from the first field (`field1` below) when part of `RolloverInfo` e.g. `"auto_sharding": { "field1" ... }` vs outside the object when parsed independently  - the entire parsing code will be part of the rollover condition and not rely necessarily on this object encapsulation. So removing the Writeable/XContent stuff in this PR as will be different in the subsequent PR where it's actually used)